File: Dinner/Java/Dinner/src/Dijkstra.java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Dijkstra extends Thread {

    private static final int FILOSOFOS = 5;
    private static final int ITERACOES = 100; // Número de vezes que cada filósofo vai comer
    private static final Lock[] garfos = new Lock[FILOSOFOS];
    private static final int[] contagemComidas = new int[FILOSOFOS];
    private static final Thread[] filosofosThreads = new Thread[FILOSOFOS];

    public static void main(String[] args) {
        for (int i = 0; i < FILOSOFOS; i++) {
            garfos[i] = new ReentrantLock();
        }

        long tempoInicio = System.currentTimeMillis();

        for (int i = 0; i < FILOSOFOS; i++) {
            final int id = i;
            filosofosThreads[i] = new Thread(() -> {
                for (int j = 0; j < ITERACOES; j++) {
                    System.out.println(id + " senta");
                    if (id % 2 == 0) {
                        garfos[id].lock();
                        System.out.println(id + " pegou o primeiro garfo");
                        garfos[(id + 1) % FILOSOFOS].lock();
                        System.out.println(id + " pegou o segundo garfo");
                    } else {
                        garfos[(id + 1) % FILOSOFOS].lock();
                        System.out.println(id + " pegou o segundo garfo");
                        garfos[id].lock();
                        System.out.println(id + " pegou o primeiro garfo");
                    }

                    contagemComidas[id]++; // Aumenta a contagem de comidas
                    garfos[id].unlock();
                    garfos[(id + 1) % FILOSOFOS].unlock();
                    System.out.println(id + " liberou os garfos");
                }
            });
            filosofosThreads[i].start(); // Inicia a thread do filósofo
        }

        // Aguardar o término dos filósofos
        for (int i = 0; i < FILOSOFOS; i++) {
            try {
                filosofosThreads[i].join(); // Espera a thread do filósofo terminar
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        long tempoFim = System.currentTimeMillis();
        long tempoTotal = tempoFim - tempoInicio;

        // Mostrar resultados
        for (int i = 0; i < FILOSOFOS; i++) {
            System.out.println("Filósofo " + i + " comeu " + contagemComidas[i] + " vezes.");
        }

        System.out.println("Tempo total de execução: " + tempoTotal + " ms");
    }
}

File: Dinner/C/dijkstra.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#define FILOSOFOS 5
#define ITERACOES 100

pthread_mutex_t garfos[FILOSOFOS];
int contagemComidas[FILOSOFOS] = {0};

void* filosofo(void* arg) {
    int id = *(int*)arg;
    for (int i = 0; i < ITERACOES; i++) {
        printf("Filósofo %d senta\n", id);
        if (id % 2 == 0) {
            pthread_mutex_lock(&garfos[id]);
            printf("Filósofo %d pegou o primeiro garfo\n", id);
            pthread_mutex_lock(&garfos[(id + 1) % FILOSOFOS]);
            printf("Filósofo %d pegou o segundo garfo\n", id);
        } else {
            pthread_mutex_lock(&garfos[(id + 1) % FILOSOFOS]);
            printf("Filósofo %d pegou o segundo garfo\n", id);
            pthread_mutex_lock(&garfos[id]);
            printf("Filósofo %d pegou o primeiro garfo\n", id);
        }

        contagemComidas[id]++;
        pthread_mutex_unlock(&garfos[id]);
        pthread_mutex_unlock(&garfos[(id + 1) % FILOSOFOS]);
        printf("Filósofo %d liberou os garfos\n", id);
    }
    return NULL;
}

int main() {
    pthread_t filosofos[FILOSOFOS];
    int ids[FILOSOFOS];
    
    for (int i = 0; i < FILOSOFOS; i++) {
        pthread_mutex_init(&garfos[i], NULL);
        ids[i] = i;
    }

    for (int i = 0; i < FILOSOFOS; i++) {
        pthread_create(&filosofos[i], NULL, filosofo, &ids[i]);
    }

    for (int i = 0; i < FILOSOFOS; i++) {
        pthread_join(filosofos[i], NULL);
    }

    for (int i = 0; i < FILOSOFOS; i++) {
        printf("Filósofo %d comeu %d vezes.\n", i, contagemComidas[i]);
    }

    for (int i = 0; i < FILOSOFOS; i++) {
        pthread_mutex_destroy(&garfos[i]);
    }

    return 0;
}
